File: dijkstra.py
class Edge:
    def __init__(self, dest, weight):
        self.dest = dest - 1
        self.weight = weight

class Node:
    def __init__(self, edges, index):
        self.edges = edges
        self.index = index-1
        self.visited = False
        self.weight = float('inf')

class Dijkstra:
    def __init__(self, nodes, start, end):
        self.nodes = nodes
        self.start = start - 1
        self.end = end - 1

    def run(self):
        nodes = self.nodes
        nodes[self.start].weight = 0
        shortest_path = float('inf')
        running = True
        while running:
            running = False
            # Iterate over the edges of unvisited nodes
            for node in nodes:
                if not node.visited and node.index >= self.start:
                    for edge in node.edges:
                        if node.weight + edge.weight < nodes[edge.dest].weight:
                            nodes[edge.dest].weight = node.weight + edge.weight
                        if edge.dest == self.end:
                            shortest_path = nodes[edge.dest].weight

                    node.visited = True
                    running = True

        return shortest_path

node1 = [Edge(2, 5),
         Edge(3, 5),
         Edge(4, 5)]

node2 = [Edge(5, 3),
         Edge(3, 2),
         Edge(4, 9)]

node3 = [Edge(4, 8)]

node4 = [Edge(5, 1)]

node5 = []

nodes = [Node(node1, 1),
         Node(node2, 2),
         Node(node3, 3),
         Node(node4, 4),
         Node(node5, 5)]

start = 1
end = 5
shortestPath = Dijkstra(nodes, start, end).run()
print(f"the shortest path from N{start} to N{end} is: {shortestPath}")

File: dijkstra.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_NODES 5
#define MAX_EDGES_PER_NODE 10

struct Edge {
    int dest;
    int weight;
};

struct Node {
    int index;
    int visited;
    int accum;
    int current_edges_num;
    struct Edge* edges;
};

struct Node* initNode(struct Node* node, int index) {
    node->index = index-1;
    node->visited = 0;
    node->accum = 9999999;
    node->current_edges_num = 0;
    node->edges = malloc(sizeof(struct Edge)*MAX_EDGES_PER_NODE);
    return node;
}

void addEdge(struct Node* node, int dest, int weight) {
    node->current_edges_num++;
    node->edges = realloc(node->edges, sizeof(struct Edge) * node->current_edges_num);
    node->edges[node->current_edges_num-1].dest = dest - 1;
    node->edges[node->current_edges_num-1].weight = weight;
}

struct Dijkstra {
    struct Node* nodes[MAX_NODES];
    int start;
    int end;
};

void initDijkstra(struct Dijkstra* dijkstra, struct Node* nodes[], int start, int end) {
    dijkstra -> start = start-1;
    dijkstra -> end = end-1;
    for (int i = 0; i < MAX_NODES; i++) {
        dijkstra -> nodes[i] = nodes[i];
    }
};

void shortest_path(struct Dijkstra* dijk){
    int shortestPath = 0;
    dijk->nodes[dijk->start]->accum=0;
    int run = 1;
    while (run==1){
        run = 0;
        for (int i = 0; i < MAX_NODES; i++) 
        {   
            if (dijk->nodes[i]->visited == 0 && dijk->nodes[i]->current_edges_num > 0) {
                for (int j = 0; j < dijk->nodes[i]->current_edges_num; j++) {
                    if (dijk->nodes[i]->edges[j].weight + dijk->nodes[i]->accum < dijk->nodes[dijk->nodes[i]->edges[j].dest]->accum) {
                        dijk->nodes[dijk->nodes[i]->edges[j].dest]->accum = dijk->nodes[i]->edges[j].weight + dijk->nodes[i]->accum;
                    }
                }
                dijk->nodes[i]->visited = 1;
                run = 1;
            }
        }
    }
    shortestPath = dijk->nodes[dijk->end]->accum;
    printf("%d\n", shortestPath);
}

int main() {

    struct Node* nodes[MAX_NODES];
    for (int i = 0; i < MAX_NODES; i++) {
        struct Node* node = malloc(sizeof(struct Node));
        nodes[i] = initNode(node, i+1);
    }
    addEdge(nodes[0], 3, 5);
    addEdge(nodes[0], 2, 4);
    addEdge(nodes[1], 5, 10);
    addEdge(nodes[2], 2, 7);
    addEdge(nodes[2], 4, 13);
    addEdge(nodes[2], 5, 2);
    addEdge(nodes[4], 4, 9);

    struct Dijkstra* dijkstra = malloc(sizeof(struct Dijkstra));
    initDijkstra(dijkstra, nodes, 1, 4);
    shortest_path(dijkstra);

    for (int i =  0; i < MAX_NODES; i++){
        free(nodes[i]->edges);  
    }

    // then free nodes
    for (int i = 0; i < MAX_NODES; i++) {
        free(nodes[i]);
    }

    free(dijkstra);
    return 0;
}